README.md

<p align="center">

<a href="https://github.com/frahik/BMTME">
<img src="Logo.png" alt="BMTME Logo"/> </a>

<h3 align="center">

Bayesian Multi-Trait Multi-Environment | Development version 1.0.19

</h4>

<p align="center">

<a href="https://www.tidyverse.org/lifecycle/#stable">
<img src="https://img.shields.io/badge/lifecycle-stable-green.svg" alt="Maturing">
</a> <a href="https://www.gnu.org/licenses/lgpl-3.0">
<img src="https://img.shields.io/badge/License-LGPL%20v3-blue.svg" alt="LGPL, Version 3.0">
</a> <a href="http://www.repostatus.org/#inactive">
<img src="https://www.repostatus.org/badges/latest/inactive.svg" alt="The project has reached a stable, usable state and is being actively developed.">
</a> <a href="https://cran.r-project.org/package=BMTME">
<img src="http://cranlogs.r-pkg.org/badges/BMTME" alt="Dowloads from the CRAN">
</a>

</p>

<p align="center">

<a href="https://travis-ci.org/frahik/BMTME">
<img src="https://travis-ci.org/frahik/BMTME.svg?branch=master" alt="Travis build status">
</a> <a href="https://ci.appveyor.com/project/frahik/bmtme">
<img src="https://ci.appveyor.com/api/projects/status/r2vog9sa30yutl5n?svg=true" alt="Appveyor build status">
</a>
<a href="https://github.com/frahik/BMTME/actions?query=workflow%3AR-CMD-check">
<img src="https://github.com/frahik/BMTME/workflows/R-CMD-check/badge.svg?branch=master" alt="Github Actions">
</a> <a href='https://coveralls.io/github/frahik/BMTME'>
<img src='https://coveralls.io/repos/github/frahik/BMTME/badge.svg?branch=master' alt='Coverage Status'/>
</a> <a href="https://cran.r-project.org/package=BMTME">
<img src="http://www.r-pkg.org/badges/version-ago/BMTME" alt="CRAN">
</a> <a href="https://cran.r-project.org/package=BMTME">
<img src="https://cranchecks.info/badges/summary/BMTME" alt="CRAN"> </a>

</p>

<h4 align="center">

\[Last README update: 2020-08-26\]

</h4>

</p>

-----

# Table of contents

  - [NEWS](#news)
      - [Instructions](#instructions)
          - [Installation](#install)
          - [Dataset](#data)
      - [How to cite this package](#cite)
      - [Contributions](#contributions)
      - [Authors](#authors)

<h2 id="news">

News of this version (1.0.19)

</h2>

See the last updates on [NEWS](NEWS.md).

<h2 id="instructions">

Instructions for proper implementation

</h2>

<h3 id="install">

Installation

</h3>

To complete installation of dev version of BMTME from GitHub, you have
to install [Rtools
Software](https://cran.r-project.org/bin/windows/Rtools/) and a few
packages first.

``` r
install.packages('devtools')
devtools::install_github('frahik/BMTME')
```

<h3 id="data">

Datasets

</h3>

The package include 6 sample datasets

| Name         | Lines | Environment | Traits | Total of observations | ME models | MTME models |
| ------------ | ----- | ----------- | ------ | --------------------- | --------- | ----------- |
| MaizeToy     | 30    | 3           | 3      | 270                   | \*        | \*          |
| WheatIranian | 30    | 2           | 2      | 120                   | \*        | \*          |
| WheatJapa30  | 30    | 1           | 6      | 180                   | \*        |             |
| WheatJapa50  | 50    | 1           | 3      | 150                   | \*        |             |
| WheatMadaToy | 50    | 1           | 6      | 300                   | \*        |             |
| WheatToy     | 30    | 3           | 2      | 180                   | \*        | \*          |

To load one dataset, use the function `data(datasetName)`

<h2 id="cite">

How to cite this package

</h2>

First option, by the article paper

    @article {Montesinos-L{\'o}pez1355,
        author = {Montesinos-L{\'o}pez, Osval A. and Montesinos-L{\'o}pez, Abelardo and Luna-V{\'a}zquez, Francisco Javier and Toledo, Fernando H. and P{\'e}rez-Rodr{\'\i}guez, Paulino and Lillemo, Morten and Crossa, Jos{\'e}},
        title = {An R Package for Bayesian Analysis of Multi-environment and Multi-trait Multi-environment Data for Genome-Based Prediction},
        volume = {9},
        number = {5},
        pages = {1355--1369},
        year = {2019},
        doi = {10.1534/g3.119.400126},
        publisher = {G3: Genes, Genomes, Genetics},
        URL = {https://www.g3journal.org/content/9/5/1355},
        eprint = {https://www.g3journal.org/content/9/5/1355.full.pdf},
        journal = {G3: Genes, Genomes, Genetics}
    }

Second option, by the manual package

``` r
citation('BMTME')
```

    ## 
    ## To cite package 'BMTME' in publications use:
    ## 
    ##   Francisco Javier Luna-Vazquez, Fernando H. Toledo, Osval Antonio
    ##   Montesinos-Lopez, Abelardo Montesinos-Lopez and Jose Crossa (2020).
    ##   BMTME: Bayesian Multi-Trait Multi-Environment for Genomic Selection
    ##   Analysis. R package version 1.0.19. https://github.com/frahik/BMTME
    ## 
    ## A BibTeX entry for LaTeX users is
    ## 
    ##   @Manual{,
    ##     title = {BMTME: Bayesian Multi-Trait Multi-Environment for Genomic Selection Analysis},
    ##     author = {Francisco Javier Luna-Vazquez and Fernando H. Toledo and Osval Antonio Montesinos-Lopez and Abelardo Montesinos-Lopez and Jose Crossa},
    ##     year = {2020},
    ##     note = {R package version 1.0.19},
    ##     url = {https://github.com/frahik/BMTME},
    ##   }

<h2 id="contributions">

Contributions

</h2>

If you have any suggestions or feedback, I would love to hear about it.
Feel free to report new issues in [this
link](https://github.com/frahik/BMTME/issues/new), also if you want to
request a feature/report a bug, or make a pull request if you can
contribute.

<h2 id="research">

Research and implementations

</h2>

  - [An R Package for Bayesian Analysis of Multi-environment and
    Multi-trait Multi-environment Data for Genome-Gased
    Prediction](https://www.g3journal.org/content/9/5/1355)
  - [A Bayesian Multiple-Trait and Multiple-Environment Model Using the
    Matrix Normal Distribution](http://doi.org/10.5772/intechopen.71521)

<h2 id="authors">

Authors

</h2>

  - Francisco Javier Luna-Vázquez (Author, Maintainer)
  - Osval Antonio Montesinos-López (Author)
  - Abelardo Montesinos-López (Author)
  - José Crossa (Author)
  - Fernando Tolero (Author)


NEWS.md
BMTME v1.0.19
==============

Changes:

 * [FIX] Rcpp bug to run test


BMTME v1.0.17
==============

Changes:

 * [FIX] Rcpp bug to run test
 * [FIX] test of the package
 * [IMP] Random partition Environment and Traits now are exported as character, not as factor.
 
BMTME v1.0.12
==============

Changes:

 * [FIX] Users can overwrite the package compiler.

BMTME v1.0.11
==============

Changes:

 * [FIX] Users can overwrite the package compiler.

BMTME v1.0.10
==============

Changes:

 * [FIX] clang: warnings
 * [FIX] Deprecated `funs` from dplyr

BMTME v1.0.6
==============

Changes:

 * [Fix bug] Fixed eigen values in `BME` and `BMTME` functions.

BMTME v1.0.5
==============

Changes:

 * [Update] Fixed package to keep it compatible with new version of dplyr.
 * [Update] New parameter `tolerance` in cholesky function.


BMTME v1.0.4  [First CRAN release]
==============

Changes:

 * [Update] Examples provided to BMORS, BME, BMTME and BMORS_Env functions.
 * [Update] Description of the package.

BMTME v1.0.3
==============

Changes:

 * [Update] New tests was added.
 * [refactoring code] Update validation of cores to use only the 50% of the available cores.
 * [refactoring code] Change to snow and doSNOW packages to parallel work.
 * [refactoring code] Bug fixed.

BMTME v1.0.2
==============

Changes:

 * [Update] New tests was added.
 * [refactoring code] Now the progress bar with parallel work it is with pbapply package
 * [refactoring code] Change snow and doSNOW packages to parallel package
 * [refactoring code] New function to crossvalidation, easier to maintain.
 * [Update] Add a good description.
 

BMTME v1.0.1
==============

Changes:
 
 * [Fixed] Makevars.win to compile C++ code.
 * [refactoring code] cholesky function.


BMTME v1.0.0
==============

Changes:
 * Fixed typo
 * Include vignettes
 * Removed unused code
 
BMTME v0.0.26
==============

Changes:
 * Improvements in tests
 * Minor fixes
 

BMTME v0.0.25
==============

Changes:
 
 * Include coverage test
 * Typo fixed
 * Fixed bug in `BMTME()` function with a vector as testingSet parameter.
 * Fixed bug in `RandomPartition()` function.
 * Implemented predictions with not crossvalidation in `BMORS()` function.
 * Include `print()` function to `BMORS` object with not crossvalidation.
 

BMTME v0.0.24
==============

Changes:

 * Removed stratified random partitions.
 * Fixed barplot for BMORS_Env.
 * Good practice with <-
 * complete F or T, logical arguments.
 * Bug fixed: changed 1:length to seq_len to avoid bugs.
 * Include ORCID.

BMTME v0.0.23
==============

Changes:

 * Include more tests.
 * Fixed print message of `BMORS()` and `BMORS_Env()` functions.
 * Fixed parallel work of `BMORS()` function.
 
BMTME v0.0.22
==============

Changes:

 * Important updates to the documentation.
 * Fixed MAC-OS 10.11 El Capitan, installation (issue #1).
 
BMTME v0.0.21
==============

Changes:

 * `BMTERS()` function now is `BMORS()` function.
 * `BMTERS_Env()` function now is `BMORS_Env()` function.

BMTME v0.0.20
==============

Changes:

 * Update Windows c++ support.
 * Fixed issues #1 (Installation in Windows) and #2 (Typos). Thanks to @volpatoo and @j450h1 by reporting it.

BMTME v0.0.19
==============

Changes:
 
 * Update plot and boxplot functions
 
BMTME v0.0.18
==============

Changes:
 
 * New `predictor_Sec_complete` parameter in `BMTMERS_Env()` function.
 
BMTME v0.0.17
==============

Changes:

 * New `predictor_Sec_complete` parameter in `BMTMERS()` function.

BMTME v0.0.16
==============

Changes:

 * The MSEP was changed to MAAPE for the error estimations of the predictions.
 * Minor fixes in the documentation.
 * Now the boxplots can be ordered by the MAAPE.
 * fixed the predicted output.
 
BMTME v0.0.15
==============

Changes:
 
 * Implement parallel mode in the BMTMERS function
 * Implement validation to `parallelCores` parameter in the functions that could use it. 
 * export `n_cores` used to fit the models that could use `parallelCores` parameter.
 * fixed class of `BME` function.
 
BMTME v0.0.14
==============

Changes:

 * Update Unix support
 
BMTME v0.0.13
==============

Changes:

 * Initial development is in progress, but there has not yet been a stable, usable release suitable for the public; this is a pre-release, be careful.

BMTME v0.0.1
==============

Changes:

* Initial commit


R/BME.R
#' Bayesian Multi-Environment Model (BME)
#'
#' @param Y \code{(matrix)} Phenotypic response where each column is a different environment.
#' @param Z1 \code{(matrix)} Matrix design for the genetic effects.
#' @param nIter \code{(integer)} Number of iterations to fit the model.
#' @param burnIn \code{(integer)} Number of items to burn at the beginning of the model.
#' @param thin \code{(integer)} Number of items to thin the model.
#' @param bs \code{(integer)} Number of groups.
#' @param parallelCores \code{(integer)} Number of cores to use.
#' @param digits \code{(integer)} Number of digits of accuracy in the results.
#' @param progressBar \code{(Logical)} Show the progress bar.
#' @param testingSet \code{(object or vector)} Crossvalidation object or vector with the positions to use like testing in a cross-validation test.
#'
#' @return If the testingSet is NULL, the function returns the predictions.
#'
#' Else, if the testingSet is not NULL, the function returns the correlation of the predictions of the cross-validation test.
#' @examples
#' \donttest{
#'   data("WheatMadaToy")
#'   phenoMada <- (phenoMada[order(phenoMada$GID),])
#'
#'   #Matrix design
#'   LG <- cholesky(genoMada)
#'   ZG <- model.matrix(~0 + as.factor(phenoMada$GID))
#'   Z.G <- ZG %*% LG
#'
#'   #Pheno data
#'   Y <- as.matrix(phenoMada[, -c(1)])
#'   # Check fitting
#'   fm <- BME(Y = Y, Z1 = Z.G, nIter = 10000, burnIn = 5000, thin = 2, bs = 50)
#'
#'   # Check predictive capacities of the model with CrossValidation object
#'   pheno <- data.frame(GID = phenoMada[, 1], Env = '', Response = phenoMada[, 3])
#'   CrossV <- CV.RandomPart(pheno, NPartitions = 4, PTesting = 0.2, set_seed = 123)
#'
#'   pm <- BME(Y = Y, Z1 = Z.G, nIter = 10000, burnIn = 5000, thin = 2, bs = 50, testingSet = CrossV)
#' }
#'
#' @references
#' Montesinos-Lopez, O.A., Montesinos-Lopez, A., Crossa, J., Toledo, F.H., Perez-Hernandez, O., Eskridge, K.M., … Rutkoski, J. (2016).
#' A Genomic Bayesian Multi-trait and Multi-environment Model. G3: Genes|Genomes|Genetics, 6(9), 2725–2744. \url{https://doi.org/10.1534/g3.116.032359}.
#'
#' @importFrom stats lm rnorm var vcov
#' @importFrom tidyr gather
#' @importFrom foreach %dopar%
#' @useDynLib BMTME
#' @export
BME <- function(Y, Z1, nIter = 1000L, burnIn = 300L, thin = 2L, bs = ceiling(dim(Z1)[2]/6), parallelCores = 1, digits = 4, progressBar = TRUE, testingSet = NULL) {
  time.init <- proc.time()[3]
  parallelCores <- validate.parallelCores(parallelCores)
  if (is.null(testingSet)) {
    out <- coreME(Y, Z1, nIter, burnIn, thin, bs, digits, progressBar, testingSet)
    class(out) <- 'BME'
  } else if (parallelCores <= 1 && inherits(testingSet, 'CrossValidation')) {
    results <- data.frame()
    nCV <- length(testingSet$CrossValidation_list)
    pb <- getProgressBar('Fitting Cross-Validation :what  [:bar] :percent;  Time elapsed: :elapsed', nCV)

    for (actual_CV in seq_len(nCV)) {
      if (progressBar) {
        pb$tick(tokens = list(what = paste0(actual_CV, ' out of ', nCV)))
      }

      positionTST <- testingSet$CrossValidation_list[[actual_CV]]

      results <- rbind(results, CVME(Y, Z1, nIter, burnIn, thin, bs, digits,
                                     positionTST, actual_CV,
                                     testingSet$Environments[positionTST]))
    }

    out <- list(results = results, n_cores = parallelCores, nIter = nIter,
                burnIn = burnIn, thin = thin, executionTime = proc.time()[3] - time.init)
    class(out) <- 'BMECV'

  } else if (parallelCores > 1 && inherits(testingSet, 'CrossValidation')) {
    cl <- snow::makeCluster(parallelCores)
    doSNOW::registerDoSNOW(cl)
    nCV <- length(testingSet$CrossValidation_list)

    progress <- NULL
    if (progressBar) {
      pb <- utils::txtProgressBar(max = nCV, style = 3)
      progress <- function(n) utils::setTxtProgressBar(pb, n)
    }
    opts <- list(progress = progress)
    results <- foreach::foreach(actual_CV = seq_len(nCV), .combine = rbind, .packages = 'BMTME', .options.snow = opts) %dopar% {
      CVME(Y, Z1, nIter, burnIn, thin, bs, digits,
           testingSet$CrossValidation_list[[actual_CV]], actual_CV,
           testingSet$Environments[testingSet$CrossValidation_list[[actual_CV]]])
    }

    parallel::stopCluster(cl)
    out <- list(results = results, n_cores = parallelCores, nIter = nIter,
                burnIn = burnIn, thin = thin, executionTime = proc.time()[3] - time.init)
    class(out) <- 'BMECV'
  } else {
    results <- CVME(Y, Z1, nIter, burnIn, thin, bs, digits, testingSet, 1, NA)

    out <- list(results = results, n_cores = parallelCores, nIter = nIter,
                burnIn = burnIn, thin = thin,
                executionTime = proc.time()[3] - time.init)
    class(out) <- 'BMECV'
  }

  return(out)
}

CVME <- function(Y, Z1, nIter, burnIn, thin, bs, digits, testingSet, iterationNumber, Env){
  fm <- coreME(Y, Z1, nIter, burnIn, thin, bs, digits, FALSE, testingSet)
  observed <- tidyr::gather(as.data.frame(Y[testingSet, ]), 'Trait', 'Observed')
  predicted <- tidyr::gather(as.data.frame(fm$yHat[testingSet, ]), 'Trait', 'Predicted')

  return(data.frame(Position = testingSet,
                    Environment = Env,
                    Trait = rep(colnames(Y), each = length(testingSet)),
                    Partition = iterationNumber,
                    Observed = round(observed$Observed, digits),
                    Predicted = round(predicted$Predicted, digits)))
}

coreME <- function(Y, Z1, nIter, burnIn, thin, bs, digits, progressBar, testingSet){
  Y[testingSet, ] <- NA

  if ((nIter - burnIn - thin) < 0L) {
    stop("nIter must be greater than thin+burnIn")
  }

  pb <- progress::progress_bar$new(format = "Fitting the model [:bar]; Time remaining: :eta",
                                   total = nIter/20L, clear = FALSE, show_after = 0L)
  # ps: blocks size
  rmv_f <- function(ps, c, A, x) {
    A    <- (A + t(A))/2
    p    <- dim(A)[1L]
    k    <- floor(p / ps)#Numbers of blocks
    r1   <- p - k * ps
    ps_1 <- ps + r1 %/% k
    r2   <- p - k * ps_1
    ps_2 <- ps_1 + 1L
    k1   <- k - r2
    tmp  <- 0L
    for (i in seq_len(k1)) {
      tmp1  <- tmp + ps_1
      Pos_i <- (tmp + 1L):tmp1
      tmp   <- tmp1
      A_ii  <- A[Pos_i, Pos_i]
      EigenA <- eigen(A_ii)
      d_A   <- EigenA$values
      V_A   <- EigenA$vectors
      pos_A1 <- which(Re(d_A) > 1e-10)
      if (identical(pos_A1, integer(0))) {
        pos_A <- 1L
      } else {
        pos_A <- pos_A1
      }
      d_A_Star <- d_A[pos_A]
      V_A_Star <- V_A[, pos_A]

      if (length(pos_A) == 1L) {
        d_A_Star_Inv <- 1 / d_A_Star
        V_A_Star_t <- d_A_Star_Inv * t(V_A_Star)
        A_ii_inv <- V_A_Star %*% V_A_Star_t
      } else {
        d_A_Star_Inv <- diag(1 / d_A_Star)
        A_ii_inv <- MatMul(MatMul(V_A_Star, d_A_Star_Inv), t(V_A_Star))
      }
      mu_i <- c(A_ii_inv %*% (c[Pos_i] - A[Pos_i, -Pos_i] %*% x[-Pos_i]))
      x[Pos_i] <- c(mvtnorm::rmvnorm(1, mu_i, A_ii_inv))
    }
    if (r2 != 0) {
      for (i in (k1 + 1L):k) {
        tmp1 <- tmp + ps_2
        Pos_i <- (tmp + 1L):tmp1
        tmp <- tmp1
        A_ii <- A[Pos_i, Pos_i]
        EigenA <- eigen(A_ii)
        d_A <- EigenA$values
        V_A <- EigenA$vectors
        pos_A1 <- which(Re(d_A) > 1e-10)
        if (identical(pos_A1, integer(0))) {
          pos_A <- 1L
        } else {
          pos_A <- pos_A1
        }
        d_A_Star <- d_A[pos_A]
        V_A_Star <- V_A[, pos_A]
        if (length(pos_A) == 1L) {
          d_A_Star_Inv <- 1 / d_A_Star
          V_A_Star_t <- d_A_Star_Inv * t(V_A_Star)
          A_ii_inv <- V_A_Star %*% V_A_Star_t
        } else {
          d_A_Star_Inv <- diag(1 / d_A_Star)
          A_ii_inv <-
            MatMul(MatMul(V_A_Star, d_A_Star_Inv), t(V_A_Star))
        }
        mu_i <- c(A_ii_inv %*% (c[Pos_i] - A[Pos_i, -Pos_i] %*% x[-Pos_i]))
        x[Pos_i] <- c(mvtnorm::rmvnorm(1, mu_i, A_ii_inv))
      }
    }
    x
  }

  n <- nrow(Y)
  nt <- ncol(Y)
  nJ <- ncol(Z1)
  b1 <- matrix(0.1, nrow = nJ, ncol = nt, dimnames = list(NULL, colnames(Y)))
  G_invg <- diag(nJ)

  post_beta <- matrix(nrow = 1L, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_beta_2 <- matrix(nrow = 1L, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_b1 <- matrix(nrow = nJ, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_b1_2 <- matrix(nrow = nJ, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_var_b1 <- matrix(nrow = nt, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_var_b1_2 <- matrix(nrow = nt, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_var_e <- matrix(nrow = nt, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_var_e_2 <- matrix(nrow = nt, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_yHat <- matrix(nrow = n, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  post_yHat_2 <- matrix(nrow = n, ncol = nt, 0L, dimnames = list(NULL, colnames(Y)))
  # post_logLik <- 0L

  YStar <- Y
  vt <- ve <- 5L
  R2 <- 0.25
  R2e <- 0.5
  my.model <- lm(YStar ~ 1)
  beta0 <- my.model$coefficient
  Cov_Beta_Inv <- vcov(my.model)

  whichNa <- list()
  whichNa$subjects <- which(apply(FUN = any, X = is.na(Y), MARGIN = 1L))
  nNa <- length(whichNa$subjects)
  tst <- c(as.numeric(whichNa$subjects))

  X <- rep(1L, n)
  tX <- t(X)
  tZ1 <- t(Z1)
  tXX <- sum(X)
  tZ1Z1 <- MatMul(tZ1, Z1)
  u_b0 <- X %*% beta0
  u_b1 <- MatMul(Z1, b1)
  betav <- beta0

  VarY <- var(Y, na.rm = TRUE)
  St <- VarY * R2 * (vt + 2L)
  Se <- VarY * (1L - R2e) * (ve + 2L)
  sigmaT <- St / (vt + 2L)
  EigenT <- eigen(sigmaT)
  d_T <- EigenT$values
  V_T <- EigenT$vectors
  pos_T <- which(d_T > 1e-10)
  d_T_Star <- d_T[pos_T]
  V_T_Star <- V_T[, pos_T]
  sigmaT.Inv <- MatMul(MatMul(V_T_Star, diag(1 / d_T_Star)), t(V_T_Star))

  yHat <- (u_b0 + u_b1)
  YStar1 <- YStar
  YStar1[tst, ] <- rep(0L, nt)

  e <- (YStar1 - u_b0)

  Re <- (var(e, na.rm = TRUE) * (1L - R2e)) / 2
  e <- (YStar1 - u_b0)
  EigenRe <- eigen(Re)
  d_Re <- EigenRe$values
  V_Re <- EigenRe$vectors
  pos_Re <- which(d_Re > 1e-10)
  d_Re_Star <- d_Re[pos_Re]
  V_Re_Star <- V_Re[, pos_Re]
  Re.Inv <- MatMul(MatMul(V_Re_Star, diag(1 / d_Re_Star)), t(V_Re_Star))
  W <- YStar1
  nSums <- 0L

  for (t in seq_len(nIter)) {
    # logLik <- 0L
    ##### Linear predictor #####################################
    e <- e + u_b0

    ##### Sample of Betas from  a normal distribution ###########
    sigmaB.Inv <- Cov_Beta_Inv / 1E1
    M <- sigmaB.Inv + Re.Inv * tXX
    mu_ac <- MatMul(sigmaB.Inv, matrixcalc::vec(beta0)) + MatMul(Krone(Re.Inv, tX), matrixcalc::vec(e))
    betav1 <- t(MVnormvv(mu_ac, M))
    betav <- matrix(betav1, ncol = nt, byrow = FALSE)
    u_b0 <- X %*% betav
    e <- e - u_b0

    ##### Sample of b1 from a normal distribution ###############
    e <- e + u_b1
    sigmaTG.Inv <- Krone(sigmaT.Inv, G_invg)
    M1 <- sigmaTG.Inv + Krone(Re.Inv, tZ1Z1)
    mu_b1 <- MatMul(Krone(Re.Inv, tZ1), matrixcalc::vec(e))
    b11 <- rmv_f(ps = bs, c = mu_b1, A = M1, x = b1)
    b1 <- matrix(b11, ncol = nt, byrow = FALSE)
    u_b1 <- MatMul(Z1, b1)
    e <- e - u_b1

    ##### Sample of sigma_Traits#################################
    tb1b1 <- MatMul(t(b1), (G_invg %*% b1))
    sigmaT <- inv_wishart(vt + nJ + nt - 1L,  tb1b1 + St)
    EigenT <- eigen(sigmaT)
    d_T <- EigenT$values
    V_T <- EigenT$vectors
    pos_T <- which(d_T > 1e-10)
    d_T_Star <- d_T[pos_T]
    V_T_Star <- V_T[, pos_T]
    sigmaT.Inv <- MatMul(V_T_Star, MatMul(diag(1 / d_T_Star), t(V_T_Star)))

    ##### Sample of sig.e########################################
    yHat <- W - e
    Re <- inv_wishart(ve + n + nt - 1, MatMul(t(e), e) + Se)
    EigenRe <- eigen(Re)
    d_Re <- EigenRe$values
    V_Re <- EigenRe$vectors
    pos_Re <- which(d_Re > 1e-10)
    d_Re_Star <- d_Re[pos_Re]
    V_Re_Star <- V_Re[, pos_Re]
    Re.Inv <- MatMul(V_Re_Star, MatMul(diag(1 / d_Re_Star), t(V_Re_Star)))
    if ((nNa > 0L)) {
      W[tst, ] <- yHat[tst, ] + mvtnorm::rmvnorm(nNa, mean = rep(0, nt), sigma = Re, method = "chol")
      e[tst, ] <- W[tst, ] - yHat[tst, ]
    }

    ##### Saving output #########################################
    if (progressBar && t %% 20L == 0L) {
      pb$tick()
    }

    if ((t > burnIn) & (t %% thin == 0L)) {
      nSums <- nSums + 1L
      k <- (nSums - 1L) / (nSums)
      post_beta <- post_beta * k + betav / nSums
      post_beta_2 <- post_beta_2 * k + (betav ^ 2) / nSums
      post_b1 <- post_b1 * k + b1 / nSums
      post_b1_2 <- post_b1_2 * k + (b1 ^ 2) / nSums

      post_var_b1 <- post_var_b1 * k + sigmaT / nSums
      post_var_b1_2 <- post_var_b1_2 * k + (sigmaT ^ 2) / nSums

      post_var_e <- post_var_e * k + Re / nSums
      post_var_e_2 <- post_var_e_2 * k + (Re ^ 2) / nSums

      post_yHat <- post_yHat * k + yHat / nSums
      post_yHat_2 <- post_yHat_2 * k + (yHat ^ 2) / nSums

      out <- list(
        Y = round(Y, digits),
        nIter = nIter,
        burnIn = burnIn,
        thin = thin,
        dfe = ve,
        Se = Se,
        yHat = round(post_yHat, digits),
        SD.yHat = round(sqrt(post_yHat_2 - (post_yHat ^ 2)), digits),
        beta = round(post_beta, digits),
        SD.beta = round(sqrt(post_beta_2 - post_beta ^ 2), digits),
        b1 = round(post_b1, digits),
        SD.b1 = round(sqrt(post_b1_2 - post_b1 ^ 2), digits),
        vare = round(post_var_e, digits),
        SD.vare = round(sqrt(post_var_e_2 - post_var_e ^ 2), digits),
        varTrait = round(post_var_b1, digits),
        SD.varTrait = round(sqrt(post_var_b1_2 - post_var_b1 ^ 2), digits),
        NAvalues = nNa)
    }
  }
  return(out)
}
